File: frontend/src/scorer.js
const TEMPERATURE_MAX = 15
const CANVAS_MAX = 25
const LOW_CLOUD_MAX = 20
const HUMIDITY_MAX = 10
const VISIBILITY_MAX = 10
const TOTAL_CLOUD_MAX = 10
const MAX_POSITIVE_SCORE = TEMPERATURE_MAX + CANVAS_MAX + LOW_CLOUD_MAX + HUMIDITY_MAX + VISIBILITY_MAX + TOTAL_CLOUD_MAX

function findEventIndex(hourlyTimes, eventTime) {
  const target = new Date(eventTime).getTime()
  let bestIdx = 0
  let minDiff = Infinity
  hourlyTimes.forEach((t, idx) => {
    const diff = Math.abs(new Date(t).getTime() - target)
    if (diff < minDiff) {
      minDiff = diff
      bestIdx = idx
    }
  })
  return bestIdx
}

function formatEventTime(iso, utcOffsetSeconds) {
  const d = new Date(iso)
  const ms = d.getTime() + utcOffsetSeconds * 1000
  const local = new Date(ms)
  let hours = local.getUTCHours()
  const minutes = local.getUTCMinutes()
  const ampm = hours >= 12 ? 'PM' : 'AM'
  hours = hours % 12 || 12
  return `${hours}:${minutes.toString().padStart(2, '0')} ${ampm}`
}

function formatEventDate(iso) {
  const d = new Date(iso)
  return d.toLocaleDateString('en-US', { weekday: 'long', month: 'long', day: 'numeric' })
}

function buildFiveDayForecast(data, currentEventIso, event) {
  const events = data.daily?.[event] || []
  const hourly = data.hourly || {}
  const hourlyTimes = hourly.time || []
  if (!events.length || !hourlyTimes.length) return []

  let startIdx = events.indexOf(currentEventIso)
  if (startIdx < 0) startIdx = 0
  const upcoming = events.slice(startIdx + 1, startIdx + 6)

  return upcoming.map((iso) => {
    try {
      const { score } = scoreSunset(hourly, findEventIndex(hourlyTimes, iso))
      const eventDate = new Date(iso)
      return {
        day_name: eventDate.toLocaleDateString('en-US', { weekday: 'short' }),
        short_date: `${eventDate.toLocaleDateString('en-US', { month: 'short' })} ${eventDate.getDate()}`,
        event_time: formatEventTime(iso, data.utc_offset_seconds || 0),
        score,
        sunset_iso: iso,
        event,
      }
    } catch {
      return null
    }
  }).filter(Boolean)
}

export function scoreSunset(hourly, idx) {
  const notes = []
  let score = 0

  const totalCloud = hourly.cloud_cover?.[idx] ?? 0
  const midCloud = hourly.cloud_cover_mid?.[idx] ?? 0
  const highCloud = hourly.cloud_cover_high?.[idx] ?? 0
  const lowCloud = hourly.cloud_cover_low?.[idx] ?? 0
  const humidity = hourly.relative_humidity_2m?.[idx] ?? 0
  const visMeters = hourly.visibility?.[idx] ?? 0
  const visKm = visMeters / 1000
  const precipProb = hourly.precipitation_probability?.[idx] ?? 0
  const tempC = hourly.temperature_2m?.[idx] ?? 0

  const maxUpper = Math.max(midCloud, highCloud)
  const combinedCanvas = (midCloud + highCloud) / 2

  if (maxUpper >= 30 || combinedCanvas >= 25 || (midCloud + highCloud) >= 40) {
    score += 25
    if (highCloud >= 30 && midCloud < 20) {
      notes.push(`High cirrus clouds (${highCloud}%) create ideal twilight color scattering`)
    } else if (midCloud >= 30 && highCloud < 20) {
      notes.push(`Mid-level altocumulus clouds (${midCloud}%) form a rich golden-hour canvas`)
    } else {
      notes.push(`Excellent cloud canvas (${combinedCanvas.toFixed(0)}% avg) structure for vibrant colors`)
    }
  } else if (totalCloud <= 20 && visKm >= 12) {
    score += 25
    notes.push(`Crystal-clear skies with excellent visibility (${visKm.toFixed(1)} km) suit a clean gradient sunset`)
  } else if (maxUpper >= 15 || combinedCanvas >= 12) {
    score += 18
    notes.push(`Promising mid/high cloud patches (${maxUpper}% max upper cloud)`)
  } else if (combinedCanvas >= 5) {
    score += 10
    notes.push(`Light mid/high cloud cover (${combinedCanvas.toFixed(0)}%)`)
  } else {
    notes.push(`Minimal upper cloud canvas (${combinedCanvas.toFixed(0)}%)`)
  }

  if (lowCloud <= 30 && totalCloud >= 25) {
    score += 10
    notes.push(`Productive sky coverage (${totalCloud}%) without heavy low overcast`)
  } else if (totalCloud < 25 && lowCloud <= 25) {
    score += 6
    notes.push(`Sparse total cloud cover (${totalCloud}%) leaves mostly open sky`)
  } else if (lowCloud <= 50) {
    score += 4
    notes.push(`Moderate cloud cover (${totalCloud}%) with partial horizon clearance`)
  } else {
    score += 1
    notes.push(`Heavy total cloud cover (${totalCloud}%) with risk of overcast`)
  }

  if (lowCloud <= 25) {
    score += 20
    notes.push(`Light low clouds (${lowCloud}%), keeping the horizon clear for direct sun rays`)
  } else if (lowCloud <= 55) {
    score += 10
    notes.push(`Moderate low clouds (${lowCloud}%) may partially filter horizon light`)
  } else {
    notes.push(`Heavy low clouds (${lowCloud}%) likely block the horizon sunbeam`)
  }

  if (humidity >= 25 && humidity <= 55) {
    score += 10
    notes.push(`Optimal humidity (${humidity}%) for crisp, vivid twilight hues`)
  } else if (visKm >= 12 && humidity <= 82) {
    score += 8
    notes.push(`High coastal air clarity (${visKm.toFixed(1)} km vis) balances moisture (${humidity}% RH)`)
  } else if (humidity < 25) {
    score += 7
    notes.push(`Dry air (${humidity}%) offers crisp horizon visibility`)
  } else {
    const humidityScore = Math.max(0, Math.round(10 * (95 - humidity) / 40))
    score += humidityScore
    notes.push(`High atmospheric humidity (${humidity}%) may add haze`)
  }

  if (visKm >= 15) {
    score += 10
    notes.push(`Excellent atmospheric visibility (${visKm.toFixed(1)} km)`)
  } else if (visKm >= 8) {
    score += 6
    notes.push(`Moderate visibility (${visKm.toFixed(1)} km)`)
  } else {
    notes.push(`Reduced visibility (${visKm.toFixed(1)} km)`)
  }

  if (tempC >= 15 && tempC <= 25) {
    score += 15
    notes.push(`Optimal temperature (${tempC.toFixed(1)}°C) for vibrant sunset colors`)
  } else if ((tempC >= 10 && tempC < 15) || (tempC > 25 && tempC <= 30)) {
    score += 10
    notes.push(`Good temperature (${tempC.toFixed(1)}°C) for sunset color development`)
  } else {
    notes.push(`Temperature (${tempC.toFixed(1)}°C) may affect color intensity`)
  }

  let normalized = Math.round((score / MAX_POSITIVE_SCORE) * 100)

  if (precipProb >= 50) {
    normalized -= 15
    notes.push(`High precipitation chance (${precipProb}%) reduces sunset visibility`)
  } else if (precipProb >= 30) {
    normalized -= 8
    notes.push(`Possible scattered rain (${precipProb}%) may mute sunset colors`)
  }

  return { score: Math.max(0, Math.min(100, normalized)), notes }
}

export async function fetchScore(lat, lon, eventType = 'sunset', date = null) {
  const params = new URLSearchParams({
    latitude: lat,
    longitude: lon,
    hourly: 'cloud_cover,cloud_cover_low,cloud_cover_mid,cloud_cover_high,relative_humidity_2m,visibility,precipitation_probability,temperature_2m',
    daily: 'sunset,sunrise',
    timezone: 'auto',
    forecast_days: '7',
  })

  const response = await fetch(`https://api.open-meteo.com/v1/forecast?${params}`)
  if (!response.ok) throw new Error(`Weather API error: ${response.status}`)
  const data = await response.json()

  const events = data.daily?.[eventType] || []
  if (!events.length) throw new Error(`No ${eventType} data available`)

  let eventIso
  let isToday = false

  if (date) {
    const targetDate = date
    eventIso = events.find((iso) => iso.startsWith(targetDate))
    if (!eventIso) throw new Error(`No ${eventType} data available for ${targetDate}`)
  } else {
    const now = Date.now() / 1000 + (data.utc_offset_seconds || 0)
    const todayEvent = new Date(events[0]).getTime() / 1000
    isToday = now <= todayEvent + 7200
    eventIso = isToday ? events[0] : (events[1] || events[0])
  }

  const idx = findEventIndex(data.hourly.time || [], eventIso)
  const { score, notes } = scoreSunset(data.hourly, idx)

  return {
    lat,
    lon,
    event: eventType,
    event_time: formatEventTime(eventIso, data.utc_offset_seconds || 0),
    event_date: formatEventDate(eventIso),
    is_today: isToday,
    score,
    notes,
    forecast: buildFiveDayForecast(data, eventIso, eventType),
    cached: false,
  }
}

File: src/sunset_score/scorer.py
from datetime import datetime

TOTAL_CLOUD_MAX = 10
CANVAS_MAX = 25
LOW_CLOUD_MAX = 20
HUMIDITY_MAX = 10
VISIBILITY_MAX = 10
TEMPERATURE_MAX = 15
MAX_POSITIVE_SCORE = sum(
    (TOTAL_CLOUD_MAX, CANVAS_MAX, LOW_CLOUD_MAX, HUMIDITY_MAX, VISIBILITY_MAX, TEMPERATURE_MAX)
)


def find_event_index(hourly_times: list[str], event_time: str) -> int:
    """Find the hourly data index closest to the given event (sunset/sunrise) time.

    Args:
        hourly_times (list[str]): List of ISO formatted timestamp strings.
        event_time (str): ISO formatted timestamp string for sunset or sunrise.

    Returns:
        int: Index of the closest hourly timestamp.

    Raises:
        ValueError: If hourly_times is empty or timestamps cannot be parsed.
    """
    if not hourly_times:
        raise ValueError("hourly_times list cannot be empty.")

    target_dt = datetime.fromisoformat(event_time)

    best_idx = 0
    min_diff = None

    for idx, t_str in enumerate(hourly_times):
        dt = datetime.fromisoformat(t_str)
        diff = abs((dt - target_dt).total_seconds())
        if min_diff is None or diff < min_diff:
            min_diff = diff
            best_idx = idx

    return best_idx


# Alias for backward compatibility
find_sunset_index = find_event_index


def score_sunset(hourly_data: dict, idx: int) -> tuple[int, list[str]]:
    """Score sunset potential for a specific hourly forecast entry.

    Args:
        hourly_data (dict): Dictionary containing Open-Meteo hourly arrays:
            - cloud_cover
            - cloud_cover_low
            - cloud_cover_mid
            - cloud_cover_high
            - relative_humidity_2m
            - visibility (in meters)
            - precipitation_probability
        idx (int): Hourly index to score.

    Returns:
        tuple[int, list[str]]:
            - Score clamped between 0 and 100.
            - List of plain-English explanation notes.
    """
    score = 0
    notes: list[str] = []

    total_cloud = hourly_data.get("cloud_cover", [0])[idx]
    mid_cloud = hourly_data.get("cloud_cover_mid", [0])[idx]
    high_cloud = hourly_data.get("cloud_cover_high", [0])[idx]
    low_cloud = hourly_data.get("cloud_cover_low", [0])[idx]
    humidity = hourly_data.get("relative_humidity_2m", [0])[idx]
    vis_meters = hourly_data.get("visibility", [0])[idx]
    vis_km = vis_meters / 1000.0 if vis_meters >= 100 else vis_meters
    precip_prob = hourly_data.get("precipitation_probability", [0])[idx]

    # 1. Mid + High Cloud Cover "The Canvas" (High cirrus or mid altocumulus catch vivid twilight light)
    max_upper_cloud = max(mid_cloud, high_cloud)
    combined_canvas = (mid_cloud + high_cloud) / 2.0

    if max_upper_cloud >= 30 or combined_canvas >= 25 or (mid_cloud + high_cloud) >= 40:
        score += 25
        if high_cloud >= 30 and mid_cloud < 20:
            notes.append(f"High cirrus clouds ({high_cloud}%) create ideal twilight color scattering")
        elif mid_cloud >= 30 and high_cloud < 20:
            notes.append(f"Mid-level altocumulus clouds ({mid_cloud}%) form a rich golden-hour canvas")
        else:
            notes.append(f"Excellent cloud canvas ({combined_canvas:.0f}% avg) structure for vibrant colors")
    elif total_cloud <= 20 and vis_km >= 12:
        score += 25
        notes.append(f"Crystal-clear skies with excellent visibility ({vis_km:.1f} km) suit a clean gradient sunset")
    elif max_upper_cloud >= 15 or combined_canvas >= 12:
        score += 18
        notes.append(f"Promising mid/high cloud patches ({max_upper_cloud}% max upper cloud)")
    elif 5 <= combined_canvas < 12:
        score += 10
        notes.append(f"Light mid/high cloud cover ({combined_canvas:.0f}%)")
    else:
        notes.append(f"Minimal upper cloud canvas ({combined_canvas:.0f}%)")

    # 2. Total Cloud Cover vs Overcast Risk
    if low_cloud <= 30 and total_cloud >= 25:
        score += 10
        notes.append(f"Productive sky coverage ({total_cloud}%) without heavy low overcast")
    elif total_cloud < 25 and low_cloud <= 25:
        score += 6
        notes.append(f"Sparse total cloud cover ({total_cloud}%) leaves mostly open sky")
    elif low_cloud <= 50:
        score += 4
        notes.append(f"Moderate cloud cover ({total_cloud}%) with partial horizon clearance")
    else:
        score += 1
        notes.append(f"Heavy total cloud cover ({total_cloud}%) with risk of overcast")

    # 3. Low Cloud Cover "Horizon Clearance" (<=25%: +20, 25-50%: +10, >50%: +0)
    if low_cloud <= 25:
        score += 20
        notes.append(f"Light low clouds ({low_cloud}%), keeping the horizon clear for direct sun rays")
    elif 25 < low_cloud <= 55:
        score += 10
        notes.append(f"Moderate low clouds ({low_cloud}%) may partially filter horizon light")
    else:
        notes.append(f"Heavy low clouds ({low_cloud}%) likely block the horizon sunbeam")

    # 4. Relative Humidity & Coastal Clarity
    if 25 <= humidity <= 55:
        score += 10
        notes.append(f"Optimal humidity ({humidity}%) for crisp, vivid twilight hues")
    elif vis_km >= 12 and humidity <= 82:
        # Coastal / marine moisture with high visibility produces brilliant scattering!
        score += 8
        notes.append(f"High coastal air clarity ({vis_km:.1f} km vis) balances moisture ({humidity}% RH)")
    elif humidity < 25:
        score += 7
        notes.append(f"Dry air ({humidity}%) offers crisp horizon visibility")
    else:
        humidity_score = max(0, round(10 * (95 - humidity) / 40))
        score += humidity_score
        notes.append(f"High atmospheric humidity ({humidity}%) may add haze")

    # 5. Visibility (>=15km: +10, 8-15km: +6, <8km: +0)
    if vis_km >= 15:
        score += 10
        notes.append(f"Excellent atmospheric visibility ({vis_km:.1f} km)")
    elif 8 <= vis_km < 15:
        score += 6
        notes.append(f"Moderate visibility ({vis_km:.1f} km)")
    else:
        notes.append(f"Reduced visibility ({vis_km:.1f} km)")

    # 6. Temperature-based color intensity (15°C-25°C: +15, 10-15°C or 25-30°C: +10, else: +0)
    temp_c = hourly_data.get("temperature_2m", [0])[idx]
    if 15 <= temp_c <= 25:
        score += 15
        notes.append(f"Optimal temperature ({temp_c:.1f}°C) for vibrant sunset colors")
    elif (10 <= temp_c < 15) or (25 < temp_c <= 30):
        score += 10
        notes.append(f"Good temperature ({temp_c:.1f}°C) for sunset color development")
    else:
        notes.append(f"Temperature ({temp_c:.1f}°C) may affect color intensity")

    normalized_score = round(score / MAX_POSITIVE_SCORE * 100)

    # 7. Precipitation Penalty (>=50%: -15, 30-49%: -8)
    if precip_prob >= 50:
        normalized_score -= 15
        notes.append(f"High precipitation chance ({precip_prob}%) reduces sunset visibility")
    elif 30 <= precip_prob < 50:
        normalized_score -= 8
        notes.append(f"Possible scattered rain ({precip_prob}%) may mute sunset colors")

    clamped_score = max(0, min(100, normalized_score))

    return clamped_score, notes
